File: languages/js/src/entity/decrypter.ts
import * as proto from "../bridge/proto/encryption_entities";

export interface Decrypter {
  alg: proto.EncryptionAlg;
  args: DecrypterArgs;
  toProto(): proto.Decrypter;
}

export class AesDecrypter implements Decrypter {
  alg: proto.EncryptionAlg;
  args: DecrypterArgs;

  constructor(password: string) {
    this.alg = proto.EncryptionAlg.A256GCM;
    this.args = new DecrypterArgs(password);
  }

  public toProto(): proto.Decrypter {
    return proto.Decrypter.fromPartial({
      alg: this.alg,
      args: this.args.toProto()
    });
  }
}

export class RsaDecrypter implements Decrypter {
  alg: proto.EncryptionAlg;
  args: DecrypterArgs;

  constructor(private_key: string) {
    this.alg = proto.EncryptionAlg.RSA;
    this.args = new DecrypterArgs(private_key);
  }

  public toProto(): proto.Decrypter {
    return proto.Decrypter.fromPartial({
      alg: this.alg,
      args: this.args.toProto()
    });
  }
}

export class EciesDecrypter implements Decrypter {
  alg: proto.EncryptionAlg;
  args: DecrypterArgs;

  constructor(private_key: string) {
    this.alg = proto.EncryptionAlg.ECIES;
    this.args = new DecrypterArgs(private_key);
  }

  public toProto(): proto.Decrypter {
    return proto.Decrypter.fromPartial({
      alg: this.alg,
      args: this.args.toProto()
    });
  }
}

export class DecrypterArgs {
  password: string;
  constructor(password: string) {
    this.password = password;
  }

  public toProto(): proto.DecrypterArgs {
    return proto.DecrypterArgs.fromPartial({ key: this.password });
  }
}

File: languages/go/entity/decrypter.go
package entity

import "github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"

type Decrypter interface {
	ToProto() *proto.Decrypter
}

type AesDecrypter struct {
	Alg  proto.EncryptionAlg
	Args DecrypterArgs
}

func NewAesDecrypter(password string) AesDecrypter {
	return AesDecrypter{
		Alg: proto.EncryptionAlg_A256GCM,
		Args: DecrypterArgs{
			Key: password,
		},
	}
}

func (e AesDecrypter) ToProto() *proto.Decrypter {
	return &proto.Decrypter{
		Alg:  e.Alg,
		Args: e.Args.ToProto(),
	}
}

type RsaDecrypter struct {
	Alg  proto.EncryptionAlg
	Args DecrypterArgs
}

func NewRsaDecrypter(key string) RsaDecrypter {
	return RsaDecrypter{
		Alg: proto.EncryptionAlg_RSA,
		Args: DecrypterArgs{
			Key: key,
		},
	}
}

func (e RsaDecrypter) ToProto() *proto.Decrypter {
	return &proto.Decrypter{
		Alg:  e.Alg,
		Args: e.Args.ToProto(),
	}
}

type EciesDecrypter struct {
	Alg  proto.EncryptionAlg
	Args DecrypterArgs
}

func NewEciesDecrypter(key string) RsaDecrypter {
	return RsaDecrypter{
		Alg: proto.EncryptionAlg_ECIES,
		Args: DecrypterArgs{
			Key: key,
		},
	}
}

func (e EciesDecrypter) ToProto() *proto.Decrypter {
	return &proto.Decrypter{
		Alg:  e.Alg,
		Args: e.Args.ToProto(),
	}
}

type DecrypterArgs struct {
	Key string
}

func (e DecrypterArgs) ToProto() *proto.DecrypterArgs {
	return &proto.DecrypterArgs{
		Key: e.Key,
	}
}
